server/src/main/java/io/unitycatalog/server/service/credential/aws/AwsPolicyGenerator.java

```diff
@@ -0,0 +1,122 @@
+
+import java.net.URI;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class AwsPolicyGenerator {
+
+  static final List<String> SELECT_ACTIONS = List.of("s3:GetO*");
+  static final List<String> UPDATE_ACTIONS =
+    List.of("s3:GetO*", "s3:PutO*", "s3:DeleteO*", "s3:*Multipart*");
+
+  static final String POLICY_STATEMENT =
+    """
+    Version: 2012-10-17
+    Statement: []
+    """;
+
+  static final String BUCKET_STATEMENT =
+    """
+    Effect: Allow
+    Action:
+      - s3:ListBucket
+    Resource: []
+    Condition:
+      StringLike:
+        "s3:prefix": []
+    """;
+
+  static final String OPERATION_STATEMENT =
+    """
+    Effect: Allow
+    Action: []
+    Resource: []
+    """;
+
+  private static final ObjectMapper JSON_MAPPER = new ObjectMapper();
+  private static final ObjectMapper YAML_MAPPER = new ObjectMapper(new YAMLFactory());
+
+  // This can support generating a policy across multiple buckets and paths, however, the assumed role
+  // the policy is applied to for a scoped-session needs to have access across those buckets
+  @SneakyThrows
+  public static String generatePolicy(Set<CredentialContext.Privilege> privileges, List<String> locations) {
+    JsonNode policyRoot = loadYaml(POLICY_STATEMENT);
+    ArrayNode policyStatement = (ArrayNode) policyRoot.findPath("Statement");
+    JsonNode operationsStatement = loadYaml(OPERATION_STATEMENT);
+    policyStatement.add(operationsStatement);
+
+    // Add the appropriate S3 operations for the privileges requested
+    ArrayNode actions = (ArrayNode) operationsStatement.findPath("Action");
+    if (privileges.contains(CredentialContext.Privilege.UPDATE)) {
+      UPDATE_ACTIONS.forEach(actions::add);
+    } else if (privileges.contains(CredentialContext.Privilege.SELECT)) {
+      SELECT_ACTIONS.forEach(actions::add);
+    } else {
+      throw new NotAuthorizedException(
+        "Can't generate policy for unknown privileges '%s' for locations: '%s'"
+          .formatted(privileges, locations));
+    }
+
+    // Group each location by s3 bucket it's located in, then for each
+    // bucket, add the bucket arn for the listBucket and operations statements,
+    // then add each path as a conditional prefix
+    getBucketToPathsMap(locations).forEach(
+      (bucketName, paths) -> {
+        JsonNode listStatement = loadYaml(BUCKET_STATEMENT);
+        policyStatement.add(listStatement);
+
+        ArrayNode bucketResource = (ArrayNode) listStatement.findPath("Resource");
+        ArrayNode operationsResource = (ArrayNode) operationsStatement.findPath("Resource");
+        bucketResource.add("arn:aws:s3:::%s".formatted(bucketName));
+
+        ArrayNode conditionalPrefixes = (ArrayNode) listStatement.findPath("s3:prefix");
+        paths.forEach(
+          path -> {
+            String sanitizedPath = path.replaceAll("^/+", "");
```

Comment:
Other special characters in policy statements should also be sanitized to be safe. I think this includes `$`, `?`, `*`, `\`. (see [this page](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_variables.html))

Reply:
this is just removing any forward slashes at the beginning or s3 paths

Reply:
Yes, this is removing only forward slashes. Separately, the other special characters in user-provided s3 paths need to be handled to avoid giving away more access than what the user is authorized to do. It could be deferred to a separate PR (if so, can you please create an issue)?